AVFoundation-CombineTests/KVObservingPublisherTests.swift

```diff
@@ -0,0 +1,142 @@
+//
+//  KVObservingPublisherTests.swift
+//  AVFoundation-CombineTests
+//
+//  Created by József Vesza on 2020. 07. 23..
+//  Copyright © 2020. József Vesza. All rights reserved.
+//
+
+import XCTest
+import Combine
+
+@testable import AVFoundation_Combine
+
+class KVObservingPublisherTests: XCTestCase {
+    var sut: Publishers.KVObservingPublisher<Counter, Int>!
+    var counter: Counter!
+    var subscriptions = Set<AnyCancellable>()
+    
+    override func setUp() {
+        counter = Counter()
+        sut = Publishers.KVObservingPublisher(observedObject: counter, keyPath: \.count)
+    }
+    
+    override func tearDown() {
+        subscriptions = []
+    }
+    
+    func testWhenCounterIsIncremented_ItEmitsTheNewValue() {
+        // given
+        let expectation = XCTestExpectation(description: "Value should be received")
+        sut.sink { _ in
+            expectation.fulfill()
+        }
+        .store(in: &subscriptions)
+        
+        // when
+        counter.increment()
+        
+        // then
+        wait(for: [expectation], timeout: 0)
+    }
+    
+    func testWhenCounterIsIncrementedTwice_ItEmitsTwoValues() {
+        // given
+        let expectedValues = [1, 2]
+        var receivedValues: [Int] = []
+        
+        sut.sink { value in
+            receivedValues.append(value)
+        }
+        .store(in: &subscriptions)
+        
+        // when
+        counter.increment()
+        counter.increment()
+        
+        // then
+        XCTAssertEqual(receivedValues, expectedValues)
+    }
+    
+    func testWhenNoValuesAreRequested_ItEmitsNoValues() {
+        // given
+        let expectedValues: [Int] = []
+        var receivedValues: [Int] = []
+        
+        let subscriber = TestSubscriber<Int>(demand: 0) { values in
+            receivedValues = values
+        }
+        
+        sut.subscribe(subscriber)
+        
+        // when
+        (0..<5).forEach { _ in counter.increment() }
+        
+        // then
+        XCTAssertEqual(receivedValues, expectedValues)
+    }
+    
+    func testWhenValuesAreRequested_ItStartsEmittingValues() {
+        // given
+        let expectedValues: [Int] = [1, 2, 3, 4, 5]
+        var receivedValues: [Int] = []
+        
+        let subscriber = TestSubscriber<Int>(demand: 0) { values in
+            receivedValues = values
+        }
+        
+        sut.subscribe(subscriber)
+        
+        // when
+        subscriber.startRequestingValues(5)
```

Comment:
I'm having trouble understanding the true reason for `TestSubscriber.startRequestingValues(:)`. It would seem all it does is modify the `demand` property after it has been set in `init(demand:onComplete:)`. So this test can be write the same just by using:

```
let subscriber = TestSubscriber<Int>(demand: 5) { values in
    receivedValues = values
}
```
Apologies for the question, but what am i missing here?

Reply:
There is a subtle difference here: `TestSubscriber<Int>(demand: 5)` will set the initial demand right away, so when `sut.subscribe(subscriber)` is called, the Subscription will know right away that it needs to produce at least 5 values.

By starting with 0, and requesting values later, we test that the Subscription only starts the work if there's a demand. Take `PlayheadProgressPublisher` for example:
```swift
func request(_ demand: Subscribers.Demand) {
    requested += demand
    // if there's no demand, we don't start the work
    guard timeObserverToken == nil, requested > .none else { return }
    // if there's demand, we start the work
    ...
}
```

This is the exact path that's being validated. Although now that you mention it, it would be better to validate it the following way:
```swift
// subscribe
sut.subscribe(subscriber)
// send a value
counter.increment()
// request values
subscriber.startRequestingValues(5)
// send another value
counter.increment()
// verify that only the second value arrives to the subscriber
...
```

Reply:
Oh, that makes sense. So this tests is about changing the demand "on the run" and testing the publisher can handle that situation. I am a bit fuzzy on this still but it kinda makes sense. In a practical situation what would be an example of this happening?

Reply:
Although it may not directly apply to our use case with AVFoundation, but it could be that there is a heavy processing operation which emits tons of values. In this case you might want to implement a custom Subscriber that can update the demand on some condition: "do we have the memory headroom to process the next 100 values? If so, we request 100, otherwise we request 50".